scapy/compat.py

```diff
@@ -0,0 +1,140 @@
+## This file is part of Scapy
+## See http://www.secdev.org/projects/scapy for more informations
+## Copyright (C) Philippe Biondi <phil@secdev.org>
+## This program is published under a GPLv2 license
+
+"""
+Python 2 and 3 link classes.
+"""
+
+from __future__ import absolute_import
+import codecs
+
+import scapy.modules.six as six
+
+###########
+# Python3 #
+###########
+
+def cmp_to_key(mycmp):
+    # TODO remove me once all 'key=cmp_to_key(..)' has been fixed in utils6.py, automaton.py
+    """Convert a cmp= function into a key= function.
+    To use with sort()
+
+    """
+    class K(object):
+        def __init__(self, obj, *args):
+            self.obj = obj
+        def __lt__(self, other):
+            return mycmp(self.obj, other.obj) < 0
+        def __gt__(self, other):
+            return mycmp(self.obj, other.obj) > 0
+        def __eq__(self, other):
+            return mycmp(self.obj, other.obj) == 0
+        def __le__(self, other):
+            return mycmp(self.obj, other.obj) <= 0  
+        def __ge__(self, other):
+            return mycmp(self.obj, other.obj) >= 0
+        def __ne__(self, other):
+            return mycmp(self.obj, other.obj) != 0
+    return K
+
+def cmp(a, b):
+    """Old python 2 function"""
+    return (a > b) - (a < b)
+
+def orb(x):
+    """Same than orb() but handle python 3 too
+    Even if unnecessary.
+
+    """
+    if type(x) is str or type(x) is bytes:
+        return ord(x)
+    else:
+        return x
+
+def chb(x):
+    """Same than chr() but handle python 3 too
+    Even if unnecessary.
+
+    """
+    if type(x) is str or type(x) is bytes:
+        return x
+    else:
+        if hasattr(x, "__int__"):
+            return chr(int(x))
+        return chr(x)
+
+if six.PY3:
+    def raw(x):
+        """Convert a str, an int, a list of ints, a packet to bytes"""
+        if x is None:
+            return None
+        if isinstance(x, list):
+            return bytes(x)
+        if isinstance(x, int):
+            try:
+                return bytes([x])
+            except ValueError:
+                # Ignore out of range strings
+                return bytes(str(x), "utf8")
```

Comment:
OK this is dangerous!
`raw(255) == b'\xff'` while `raw(256) == b'256'`.

Also, do we need to have a specific case for `list` objects? They are handled by `bytes()` anyway.

Do we need to have an exception for `None`?

I think `raw()` should remain straightforward. Here is what we could have:
```Python
def raw(val):
    try:
        return bytes(val)
    except TypeError:
        return bytes(val, encoding="utf8")
```

And I think we could adapt the parts of the code that won't work with this. What do you think?